File: python/src/egy_names/_rules_config.py
"""Loader for the shared, cross-SDK rule config.

``data/logic_config.json`` (synced by scripts/sync-catalog.sh, same as
names.json.gz) is the single source of truth for every threshold and
rule list that used to be hardcoded per language: non-personal
surfaces, low-confidence detection, ML abstention thresholds, and the
gender/religion/role prefix-suffix rule tables. Only pure algorithms
(compound-token lookahead, first-personal-token-wins, corpus-share
tie-break) stay as code, because they cannot be expressed as data.

If the config file is missing or malformed, fall back to the values
last known correct from this session's audits, so the library never
hard-fails on a packaging mistake — but log-free, since this is a
library, not an app with a logger configured.
"""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any, Dict, FrozenSet, List, Tuple

_CONFIG_PATH = Path(__file__).parent / "data" / "logic_config.json"

_FALLBACK: Dict[str, Any] = {
    "quality": {
        "non_personal_ar": [
            "الله", "الرجل", "الرجال", "شربه", "لافندي", "لفندي", "ماء", "البيت",
        ],
        "uncertain_meaning_markers": [
            "غير واضح", "لا يوجد معنى", "غير معروف",
            "قد يكون تحريف", "تحريفاً", "تحريفًا",
        ],
        "low_confidence_share_epsilon": 0.0001,
        "kunya_exempt_prefixes": ["أبو", "ابو", "أم", "ام"],
    },
    "infer_thresholds": {
        "gender_min_p": 0.70,
        "muslim_min_p": 0.85,
        "christian_min_p": 0.90,
        "role_min_p": 0.88,
    },
    "infer_rules": {"gender": [], "religion": [], "role": []},
}

_config: Dict[str, Any] = {}


def _load() -> Dict[str, Any]:
    global _config
    if _config:
        return _config
    try:
        with open(_CONFIG_PATH, "r", encoding="utf-8") as f:
            _config = json.load(f)
    except (OSError, json.JSONDecodeError, KeyError):
        _config = _FALLBACK
    return _config


def non_personal_ar() -> FrozenSet[str]:
    return frozenset(_load().get("quality", {}).get("non_personal_ar", []))


def uncertain_meaning_markers() -> Tuple[str, ...]:
    return tuple(_load().get("quality", {}).get("uncertain_meaning_markers", []))


def low_confidence_share_epsilon() -> float:
    return float(_load().get("quality", {}).get("low_confidence_share_epsilon", 0.0001))


def kunya_exempt_prefixes() -> Tuple[str, ...]:
    return tuple(_load().get("quality", {}).get("kunya_exempt_prefixes", []))


def infer_thresholds() -> Dict[str, float]:
    return _load().get("infer_thresholds", _FALLBACK["infer_thresholds"])


def infer_rules(kind: str) -> List[Dict[str, Any]]:
    """Rule table for 'gender' | 'religion' | 'role'."""
    return _load().get("infer_rules", {}).get(kind, [])


def _as_list(v: Any) -> List[str]:
    if v is None:
        return []
    return [v] if isinstance(v, str) else list(v)


def match_rule(rule: Dict[str, Any], surface: str, normalized: str) -> bool:
    """Evaluate one rule against a token.

    prefix/suffix test the normalized form; contains tests the raw
    surface (diacritics-sensitive, matching the original hand-written
    checks this config replaced). match='all' requires every listed
    condition; default ('any') requires just one.
    """
    conditions: List[bool] = []
    for p in _as_list(rule.get("prefix")):
        conditions.append(normalized.startswith(p))
    for s in _as_list(rule.get("suffix")):
        conditions.append(normalized.endswith(s))
    for c in _as_list(rule.get("contains")):
        conditions.append(c in surface)
    if not conditions:
        return False
    return all(conditions) if rule.get("match") == "all" else any(conditions)

File: dart/egy_names/lib/src/rules_config.dart
/// Loader for the shared, cross-SDK rule config.
///
/// `data/logic_config.json` (synced by scripts/sync-catalog.sh, same as
/// names.json.gz) is the single source of truth for every threshold and
/// rule list that used to be hardcoded per language: non-personal
/// surfaces, low-confidence detection, ML abstention thresholds, and the
/// gender/religion/role prefix-suffix rule tables. Only pure algorithms
/// (compound-token lookahead, first-personal-token-wins, corpus-share
/// tie-break) stay as code, because they cannot be expressed as data.
///
/// If the config file is missing or malformed, fall back to the values
/// last known correct from this session's audits, so the library never
/// hard-fails on a packaging mistake.
library;

import 'dart:convert';
import 'dart:io';

import 'data.dart';

class InferThresholds {
  final double genderMinP;
  final double muslimMinP;
  final double christianMinP;
  final double roleMinP;

  const InferThresholds({
    required this.genderMinP,
    required this.muslimMinP,
    required this.christianMinP,
    required this.roleMinP,
  });
}

class RulesConfig {
  final Set<String> nonPersonalAr;
  final List<String> uncertainMeaningMarkers;
  final double lowConfidenceShareEpsilon;
  final List<String> kunyaExemptPrefixes;
  final InferThresholds inferThresholds;
  final Map<String, List<Map<String, dynamic>>> inferRules;

  const RulesConfig({
    required this.nonPersonalAr,
    required this.uncertainMeaningMarkers,
    required this.lowConfidenceShareEpsilon,
    required this.kunyaExemptPrefixes,
    required this.inferThresholds,
    required this.inferRules,
  });

  static const RulesConfig fallback = RulesConfig(
    nonPersonalAr: {
      'الله',
      'الرجل',
      'الرجال',
      'شربه',
      'لافندي',
      'لفندي',
      'ماء',
      'البيت',
    },
    uncertainMeaningMarkers: [
      'غير واضح',
      'لا يوجد معنى',
      'غير معروف',
      'قد يكون تحريف',
      'تحريفاً',
      'تحريفًا',
    ],
    lowConfidenceShareEpsilon: 0.0001,
    kunyaExemptPrefixes: ['أبو', 'ابو', 'أم', 'ام'],
    inferThresholds: InferThresholds(
      genderMinP: 0.70,
      muslimMinP: 0.85,
      christianMinP: 0.90,
      roleMinP: 0.88,
    ),
    inferRules: {'gender': [], 'religion': [], 'role': []},
  );
}

class RulesConfigLoader {
  static RulesConfig? _cached;

  static const _configRel = 'lib/src/data/logic_config.json';

  static RulesConfig load({String? customPath}) {
    if (_cached != null && customPath == null) {
      return _cached!;
    }

    final config = _tryLoad(customPath) ?? RulesConfig.fallback;

    if (customPath == null) {
      _cached = config;
    }
    return config;
  }

  static void clearCache() {
    _cached = null;
  }

  static RulesConfig? _tryLoad(String? customPath) {
    try {
      final path = _resolveConfigPath(customPath);
      final file = File(path);
      if (!file.existsSync()) return null;

      final jsonMap =
          jsonDecode(file.readAsStringSync()) as Map<String, dynamic>;
      final quality = jsonMap['quality'] as Map<String, dynamic>? ?? {};
      final inferThresholdsMap =
          jsonMap['infer_thresholds'] as Map<String, dynamic>? ?? {};
      final inferRulesMap = jsonMap['infer_rules'] as Map<String, dynamic>? ?? {};

      final nonPersonalAr = ((quality['non_personal_ar'] as List<dynamic>?) ?? [])
          .map((e) => e.toString())
          .toSet();
      final uncertainMarkers =
          ((quality['uncertain_meaning_markers'] as List<dynamic>?) ?? [])
              .map((e) => e.toString())
              .toList();
      final epsilon = (quality['low_confidence_share_epsilon'] as num?)
              ?.toDouble() ??
          RulesConfig.fallback.lowConfidenceShareEpsilon;
      final kunyaPrefixes =
          ((quality['kunya_exempt_prefixes'] as List<dynamic>?) ?? [])
              .map((e) => e.toString())
              .toList();

      final thresholds = InferThresholds(
        genderMinP: (inferThresholdsMap['gender_min_p'] as num?)?.toDouble() ??
            RulesConfig.fallback.inferThresholds.genderMinP,
        muslimMinP: (inferThresholdsMap['muslim_min_p'] as num?)?.toDouble() ??
            RulesConfig.fallback.inferThresholds.muslimMinP,
        christianMinP:
            (inferThresholdsMap['christian_min_p'] as num?)?.toDouble() ??
                RulesConfig.fallback.inferThresholds.christianMinP,
        roleMinP: (inferThresholdsMap['role_min_p'] as num?)?.toDouble() ??
            RulesConfig.fallback.inferThresholds.roleMinP,
      );

      final inferRules = <String, List<Map<String, dynamic>>>{};
      for (final kind in ['gender', 'religion', 'role']) {
        final rawList = inferRulesMap[kind] as List<dynamic>? ?? [];
        inferRules[kind] = rawList
            .whereType<Map<String, dynamic>>()
            .toList();
      }

      return RulesConfig(
        nonPersonalAr: nonPersonalAr.isNotEmpty
            ? nonPersonalAr
            : RulesConfig.fallback.nonPersonalAr,
        uncertainMeaningMarkers: uncertainMarkers,
        lowConfidenceShareEpsilon: epsilon,
        kunyaExemptPrefixes: kunyaPrefixes,
        inferThresholds: thresholds,
        inferRules: inferRules,
      );
    } catch (_) {
      return null;
    }
  }

  static String _resolveConfigPath(String? customPath) {
    if (customPath != null && File(customPath).existsSync()) {
      return customPath;
    }

    final fromConfig = DataLoader.resolveSiblingDataPath(_configRel);
    if (fromConfig != null) return fromConfig;

    final possiblePaths = [
      _configRel,
      'packages/egy_names/src/data/logic_config.json',
      '${Directory.current.path}/$_configRel',
      '${Directory.current.path}/dart/egy_names/$_configRel',
      Platform.script.resolve('src/data/logic_config.json').toFilePath(),
      Platform.script.resolve('../lib/src/data/logic_config.json').toFilePath(),
    ];

    for (final p in possiblePaths) {
      try {
        if (File(p).existsSync()) return p;
      } catch (_) {}
    }

    return _configRel;
  }
}
